src/box/box.cc

```diff
@@ -2605,6 +2613,17 @@ relay_get_sync_on_start(struct trigger *trigger, void *event)
 	/* Already accounted. */
 	if (bit_test(&data->collected_vclock_map, id))
 		return 0;
+
+	/*
+	 * Let box_collect_confirmed_vclock know that the data is needed.
+	 * Otherwise, it may free the data on return while this fiber
+	 * yielding inside relay_trigger_vclock_sync.
+	 */
+	++data->ref_count;
```

Comment:
Now you're incrementing/decrementing counter in too places: `box_collect_confirmed_vclock` and `relay_get_sync_on_start`. I'd rather put everything inside `relay_trigger_vclock_sync`. This way it'll be all in one place, you increment before yielding, and decrement, free memory after yielding.

Reply:
Done.

Reply:
I suppose we misunderstood each other. It's still in `relay_get_sync_on_start` and `box_collect_confirmed_vclock`, not in `relay_trigger_vclock_sync`. Moreover, now there's initialization with 1 in `data->ref_count`, which is not easy to parse.

Reply:
My bad, we cannot do that, as `data` is used after exiting from `rellay_trigger_vclock_sync` in `box_collect_confirm_vclock`